tc/core/tensor.cc

```diff
@@ -0,0 +1,123 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <version.h>
+
+#include <cstdint>
+#include <fstream>
+#include <numeric>
+#include <tuple>
+
+#include "tc/core/tensor.h"
+#include "tc/core/utils/math.h"
+
+namespace tc {
+
+namespace {
+uint64_t GetDLTensorAlignment(const DLTensor* t) {
+  return (reinterpret_cast<std::uintptr_t>(t->data) + t->byte_offset) % 256;
+}
+uint64_t GetDLTensorAlignment(const DLConstTensor* t) {
+  return (reinterpret_cast<std::uintptr_t>(t->data) + t->byte_offset) % 256;
+}
+} // namespace
+
+TensorInfo::TensorInfo(
+    DLDataType t,
+    uint64_t align,
+    const std::vector<int64_t>& sz,
+    const std::vector<int64_t>& st)
+    : dtype(t), alignment(align), shape(sz), strides(st) {}
+
+TensorInfo::TensorInfo(const DLTensor* t)
+    : dtype(t->dtype), alignment{GetDLTensorAlignment(t)} {
+  shape.reserve(t->ndim);
+  std::copy(t->shape, t->shape + t->ndim, std::back_inserter(shape));
+  if (not t->strides) {
+    return;
+  }
+  strides.reserve(t->ndim);
+  std::copy(t->strides, t->strides + t->ndim, std::back_inserter(strides));
+}
+
+TensorInfo::TensorInfo(const DLConstTensor* t)
+    : dtype(t->dtype), alignment{GetDLTensorAlignment(t)} {
+  shape.reserve(t->ndim);
```

Comment:
This is a copy-paste from `TensorInfo::TensorInfo(const DLTensor* t)`.   Can't we use templates, delegating constructors or some other mechanism to avoid copy-pasting?